src/memory_region.rs

```diff
@@ -52,68 +52,102 @@ impl fmt::Debug for MemoryRegion {
     }
 }
 
-/// Helper for translate_addr to generate errors
-fn generate_access_violation<E: UserDefinedError>(
-    vm_addr: u64,
-    len: usize,
-    access_type: &str,
-    pc: usize,
-    regions: &[MemoryRegion],
-) -> EbpfError<E> {
-    let mut regions_string = "".to_string();
-    if !regions.is_empty() {
-        regions_string = "regions:".to_string();
-        for region in regions.iter() {
-            regions_string = format!(
-                "  {} \n{:#x}-{:#x}",
-                regions_string,
-                region.addr_vm,
-                region.addr_vm + region.len - 1,
-            );
+/// Type of memory access
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum AccessType {
+    /// Read
+    Load,
+    /// Write
+    Store,
+}
+
+/// Indirection to use instead of a slice to make handling easier
+#[derive(Default)]
+pub struct MemoryMapping {
+    /// Mapped (valid) regions
+    regions: Vec<MemoryRegion>,
+}
+impl MemoryMapping {
+    /// Creates a new MemoryMapping structure
+    pub fn new() -> Self {
+        Self {
+            regions: Vec::new(),
+        }
+    }
+
+    /// Creates a new MemoryMapping structure from the given regions
+    pub fn new_from_regions(regions: &[MemoryRegion]) -> Self {
+        Self {
+            regions: regions.to_vec(),
         }
     }
-    EbpfError::AccessViolation(
-        access_type.to_string(),
-        pc + ELF_INSN_DUMP_OFFSET,
-        vm_addr,
-        len,
-        regions_string,
-    )
-}
 
-/// Given a list of regions translate from virtual machine to host address
-pub fn translate_addr<E: UserDefinedError>(
-    vm_addr: u64,
-    len: usize,
-    access_type: &str,
-    pc: usize, // TODO syscalls don't have this info
-    regions: &[MemoryRegion],
-) -> Result<u64, EbpfError<E>> {
-    let index = match regions.binary_search_by(|probe| probe.addr_vm.cmp(&vm_addr)) {
-        Ok(index) => index,
-        Err(index) => {
-            if index == 0 {
-                return Err(generate_access_violation(
-                    vm_addr,
-                    len,
-                    access_type,
-                    pc,
-                    regions,
-                ));
+    /// Adds a region
+    pub fn add_region(&mut self, region: MemoryRegion) {
+        self.regions.push(region);
+    }
+
+    /// Adds multiple regions
+    pub fn add_regions(&mut self, regions: Vec<MemoryRegion>) {
+        self.regions.extend(regions);
+    }
+
+    /// Call after the last change. Sorts regions by addr_vm for binary search
+    pub fn finalize(&mut self) {
+        self.regions.sort_by(|a, b| a.addr_vm.cmp(&b.addr_vm));
+    }
+
+    /// Given a list of regions translate from virtual machine to host address
+    pub fn translate_addr<E: UserDefinedError>(
```

Comment:
nit: If this is now called `MemoryMapper` maybe instead of `translate_addr` it should just be `map`.  Or maybe it should be `AddressTranslator`?